### PR TITLE
SONARJAVA-1447 enforce Try with resources

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/test/files/checks/TryWithResourcesCheck_no_java_version.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        FileReader fr = new FileReader(fileName);
+        BufferedReader br = new BufferedReader(fr)
+    ) { //compliant
+      return br.readLine();
+    }
+    catch (Exception e) {}
+
+
+  }
+}
```
Comment (mpaladin) [inline java-checks/src/test/files/checks/TryWithResourcesCheck_no_java_version.java]: @benzonico EOF new line.

File: java-checks/src/test/files/checks/TryWithResourcesCheck.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        FileReader fr = new FileReader(fileName);
+        BufferedReader br = new BufferedReader(fr)
+    ) { //compliant
+      return br.readLine();
+    }
+    catch (Exception e) {}
+
+
+  }
+}
```
Comment (mpaladin) [inline java-checks/src/test/files/checks/TryWithResourcesCheck.java]: @benzonico EOF new line.

File: java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    return ImmutableList.of(Tree.Kind.TRY_STATEMENT, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.TRY_STATEMENT)) {
+      withinTry.push((TryStatementTree) tree);
+    } else if (withinStandardTry() && ((NewClassTree) tree).symbolType().isSubtypeOf("java.lang.AutoCloseable") && !reported.contains(withinTry.peek())) {
+      reported.add(withinTry.peek());
+      reportIssue(withinTry.peek().tryKeyword(), "Change this \"try\" to a try-with-resources." + context.getJavaVersion().java7CompatibilityMessage());
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java]: @benzonico with the current implementation it is not clear what is the cause of the issue. We should probably create one issue per instantiated resource with a message like `Change the enclosing "try" to a try-with-resources.`

**Comment by mpaladin** [comment]
@benzonico false negative in this case:

    try (
      FileReader fr = new FileReader(fileName)
    ) { //compliant
      BufferedReader br = new BufferedReader(fr);
      return br.readLine();
    }

**Comment by benzonico** [comment]
IMO not a fp more an out of scope...

File: java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      toReport.peek().add(tree);
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(Tree.Kind.TRY_STATEMENT)) {
+      TryStatementTree tryStatementTree = withinTry.pop();
+      List<Tree> secondaryTrees = toReport.pop();
+      if(!secondaryTrees.isEmpty()) {
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java]: @benzonico formatting.

**Pull request merged**